### ISSUE TITLE
MadEvent integration (and Fortran interoperability with C/C++/CUDA)

### ISSUE THREAD

**Issue by valassi** [madgraph5/madgraph4gpu#121]
I open a new issue, which may eventually become a container of sub-issues, about madevent integration.

This probably started off with me wanting to get some numbers for our vCHEP talk, comparing different C/CUDA versions to one another over time, and eventually also to Fortran. This got me thinking a bit more on C/Fortran interoperability, for the sake of realistic throughput comparisons. Then on Monday's meeting we had some discussions with Taylor over how far we are from making any code usable in production, knowing that the present code uses MadEvent and we have a standalone framework using Rambo instead. So I started looking a bit at some of this in more detail, in particular at how the current MadEvent code works internally now, and also at Fortran/C interoperability options more generally.

The way I see it currently is that we need two lines of development:
1. A standalone R&D application, developed essentially from scratch. This is what we have been focusing so far. We have a single standalone main in C++, where we can plug both CUDA and C++ components for random numbers, phase space sampling and ME calculation. This has no overhead yet from storing events or any postprocessing, so it is best as a prototype of "the fastest one could theoretically achieve". We are also free to do huge disruptive changes, as we are starting from scratch. Here we will also add cross section calculations and event unweighting to have a full 'quasi-realistic' workfow. In this C++ application, I would also consider adding the existing Fortran ME calculation as another plugin, to provide a realistic comparison of Fortran/C++/CUDA throughputs for the ME part alone (Olivier was not convinced about this point, and Stefan is also working on a standalone Fortran main instead); this would be calling Fortran from a C/C++ application (with data structures allocated from C++ and used as references in Fortran).
2. The integration of our C++/CUDA components into the existing MadGraph/MadEvent framework, which is much more than just Fortran and is instead a complex interplay of Fortran applications (and even bash scripts) orchestrated in Python. In practice, we need to work on adiabatic changes to the existing Madgraph, to slowly inject into it a faster ME calculation in C++ and/or CUDA (and possibly more later on). I think that this is unavoidable: if we want existing users of MadGraph to use our changes, we must keep the existing user interface and user functionalities. To get there, we should rather link our C++/CUDA components into the MadEvent Fortran application. In practice, this is calling C (and/or C++/CUDA) from Fortran. It is not obvious where the data structures would be allocated (presently they are allocated in Fortran, but this may need to change). One big change that must happen here is also that the current MadEvent interfaces, which deal with one event at a time, must evolve to deal with a large bunch of events in parallel (the ~500k events processed in parallel on GPU threads, or even just the 4-8 events processed in parallel in vectorized C++). [By the way, this probably needs to stay Make-based rather than CMake-based for the moment, as the rest of MadGraph]. Note that the existing MadGraph has a large overhead, so this route cannot be used for testing "how fast we can theoretically go". 

One general idea for both lines of development should be modularity (allowing plug-and-play replacement of components). To get there, we also need to streamline/understand/document as much as possible the data flow amongst all components involved (we did progress, but there is much more to do).

This specific github issue is mainly about the second "line of development", namely madevent integration. I started looking a bit more in detail at how MadGraph/MadEvent works internally, and also a bit at how to plug C++ in there. (This is also needed for the first line of development, because we must design components for the standalone application that have the right interface to be used as-is within MadEvent later on).

I have bits and pieces of code that I will send as a PR, also useful for timing measurements of the existing MadEvent (my original starting point), and containing bits of C code injection. And (with Olivier's help) I started to study a bit more about MadGraph/MadEvent works, I will try to document this here, so we can try to get a better "big picture".

**Comment by oliviermattelaer** [comment]
Hi Andrea, 

Funny but I'm worried about completely different points:

1. The state information does not bother me, I see this as a database that need to be extended with one more index
2. My way of interfacing is to keep gen_mom, sample_get_x, ... intact and instead of computing the weight, feed the database (and potentially extend it to include momenta and other information not yet in the database) and postpose the computation of the matrix-element up to the point where the database is complete such that we will have a simple for loop where we can use openmp/gpu/vector, first for the evaluation of the weight, but potentially also for all the following steps --initial unweighting, event writting,...)

3. The only point that worries me is what we call group-subprocessing in madevent (we do not have that optimization at NLO niether at standalone). This is an optimization which makes one channel of integration to handle in the same executable multiple matrix-element. In our implementation each phase-space point only evaluates ONE of that set of matrix-element. This is an optimization that will creates us trouble for sure. For openmp/vextorization this is likely not a big issue but we will pay a bigger cost for GPU.


Cheers,

Olivier

PS: This LO optimization can be deactivated so this is not a blocker per se and I have different idea to experiment here

**Comment by valassi** [comment]
Hi Olivier, sorry for the delay and thanks a lot! It is very reassuring that you do not see a big issue :-)

Just to summarise and try to converge in some tentative concrete proposal:

1. About what I called the "stateful" issue, meaning that DSIG1 and MATRIX1 have several if/then blocks which also depend on common blocks.

If I understand correctly, you say that it should relatively easy (in the Fortran) to restructure this into a first component that fills a "database" as output, and then a second component that computes the MEs as output based on input momenta and on that "database" as additional input. A couple of observations/questions:
- Does this "database" contain global properties of the process/application, or per-event properties? If they are only global properties, while the only per-event input to the ME calculation is the momenta, then indeed it's not too bad.
- In any case, within this ME-computing second component, I would try to clearly separate a sub-component that ONLY takes the momenta as input (i.e. none of the global properties in the database, or as few as possible), so that we can write that as a pluggable piece that can be implemented in C++/CUDA (at least for a single event, to start with).
- Do you think this makes sense, e.g. could you restructure the Fortran DSIG1/MATRIX1 along these lines?
- [About the optimization you mention in your point 3, I am not sure I understand it precisely, but hopefully we can address it in similar ways, calling several ME calculations instead of only one?]

2. About single-event vs multi-event interfaces, I am afraid this remains a very important and tricky part. 

I agree with you that the easiest (at least initially) is to try and keep gen_mom, sample_get_x etc exactly as they are. However, they have all single-event interface, so they do "loop on event i; get momenta of i; compute ME of i; increase to i+1 and iterate", while for that 'compute ME of event i' we must have a whole lot of events. A few observations/questions:
- Ideally, it would be best to restructure keep gen_mom, sample_get_x etc to handle many events in parallel (as in out curand/rambo on the GPU), but this looks very very difficult, it's as bad as reimplemnting these routines from scratch.
- The only alternative I see is to accumulate momenta before calling the ME computation. In practice, in dsample.f one should split the loop after x_to_f_arg and before dsig? Instead of "loop on event i; call x_to_f_arg(i); call dsig(i); post-process(i); next i+1", this would become "loop on event i; call x_to_f_arg(i); store_somewhere_momenta(i); next i+i" and "loop on event i; read_somewhere_momenta(i); call x_to_f_arg(i); call dsig(i); post-process(i); next i+1". This temporary storage of momenta could be somewhere in memory, or even on file? Again, the idea is to store/write only the per-event info, not the global database: I imagine this would be only the per-event momenta and sampling weight?
https://github.com/madgraph5/madgraph4gpu/blob/9f2d81a7de3b8b9d03f7153f85f76b96ad6656e0/epoch1/gridpack/eemumu/madevent/Source/dsample.f#L135

3. About the event sample sizes and control flow in event loops - this is another important/tricky part.

What I suggest above assumes that all loops on events have a size that is pre-determined in advance, i,.e. that there is no change to the control flow during the loop. In partoicular:
- With respect to your "database", this means that the "database" is not updated during the event loop (or maybe, it is updated during the x_to_f_arg part of the loop, but not during the dsig/postprocessing part of the loop). At most, the only control flow decision one may expect is "I have enough events, abort the loop prematurely", but nothing else. Is this the case, or not?
- One specific observation that puzzles me a lot is what I reported in #123: the unweighting efficiency seems to vary with the number of unweighted events requested. So it's as if the global strategy and control flow was modified during the loop itself. What is it that I do not understand there? (Maybe answer in #123 please...)

Hope this helps... just some brainstorming...
Thanks!
Andrea

**Comment by valassi** [comment]
Hi @roiser, as we just discussed, this is the Fortran I think we need to modify, from the link above
```
c
c     Main Integration Loop
c
      iter = 1
      do while(iter .le. itmax)
c
c     Get integration point
c
         call sample_get_config(wgt,iter,ipole)
         if (iter .le. itmax) then
            ievent=ievent+1
            call x_to_f_arg(ndim,ipole,mincfig,maxcfig,ninvar,wgt,x,p)
            if (pass_point(p)) then
               fx = dsig(p,wgt,0) !Evaluate function
               wgt = wgt*fx
               if (wgt .ne. 0d0) call graph_point(p,wgt) !Update graphs
            else
               fx =0d0
               wgt=0d0
            endif
            call sample_put_point(wgt,x(1),iter,ipole) !Store result
         endif
         if (wgt .ne. 0d0) kevent=kevent+1    
c
c     Write out progress/histograms
c
         if (kevent .ge. nwrite) then
            nwrite = nwrite+ncall*itmax/nsteps
            nwrite = min(nwrite,ncall*itmax)
            call graph_store
         endif
 99   enddo
```

First point, I think we need to split this here
```
(event loop)
            call x_to_f_arg(ndim,ipole,mincfig,maxcfig,ninvar,wgt,x,p)
            if (pass_point(p)) then
               fx = dsig(p,wgt,0) !Evaluate function
               wgt = wgt*fx
               if (wgt .ne. 0d0) call graph_point(p,wgt) !Update graphs
            ...
            call sample_put_point(wgt,x(1),iter,ipole) !Store result
```
This should become two loops, where we replace dsig
```
(event loop)
            call x_to_f_arg(ndim,ipole,mincfig,maxcfig,ninvar,wgt,x,p)
            if (pass_point(p)) then
            => "store point p into a big PArray and point wgt into a big WgtArray"
(inject c++/cuda and replace dsig(p,wgt,0) and )
            => "pass big PArray and WgtArray to CUDA/C++ and get back a big FXarray, or actually a big WgtArray"
            => in the above, also compute wgt = wgt*fx as an array
(event loop)
            => iterate over WgtArray one by one
            ...
            call sample_put_point(wgt,x(1),iter,ipole) !Store result
```

Second point, this assumes that fx = dsig(p,wgt,0) only computes fx (maybe also wgt?), and that there are no commons there. I mean, one should clarify the inputs and outputs.

About the big arrays, I would probably create them both (input and output) in the C++, as plain C arrays. So essentially you need three C wrappers to call from Fortran: one wrapper that stores individual p's, one wrapper that computes array w from array p, and one wrapper that extracts the next wgt from the wgt array. PLus probably two that do initialisation (malloc of the arrays) and on ethat does finalistaion (free).

About the interfaces, this is the file where I did very somple interfaing, with trailing underscores
https://github.com/madgraph5/madgraph4gpu/blob/9f2d81a7de3b8b9d03f7153f85f76b96ad6656e0/epoch1/gridpack/eemumu/madevent/SubProcesses/P1_ll_ll/counters.cpp#L9
My examples was simple as I have void functions of void. So the next step here is to test the passing of C style arrays between the fortran and C.

On the C++/cuda side, what needs to be done is to clean up the interfaces of CPPProcess to encapsulate all that is needed, without relying on check.cc. I mean I would put the creation of arrays, and a lot of the parametsr in mgOnGpuConfig into CPPPProcess, decoupling completely rambo and random numbers. Related to #175 and #176, but slightly different actually.

**Comment by roiser** [comment]
Hi @valassi I did play a bit with the passing arrays from fortran to c++ and back. It's really easy to do. I managed to pass arrays of int, float, double, complex<float> and complex<double> from fortran to c++, modify them there and then pass back. The data structures map without problems so I propose we have all defined on the fortran side. See attached 
[example.zip](https://github.com/madgraph5/madgraph4gpu/files/6399880/example.zip)

**Comment by valassi** [comment]
Interesting point. Do you mean that we need to compute more than one calculation for different values of alphaQCD for each event, or that we need to somehow calculate the appropriate value of alphaQCD for that event kinematics? (Sorry, my physics is a bit rusted...).

If it's just about extra values of alphaQCD, no, I would keep the parallelism at the level of events (the _sv thing). Essentially it's another calculation. But we should do it at the same time, so we do not need to refetech th emomenta from memory. Just some ideas...

**Comment by valassi** [comment]
> 
> 
> Hi @valassi I did play a bit with the passing arrays from fortran to c++ and back. It's really easy to do. I managed to pass arrays of int, float, double, complex<float> and complex<double> from fortran to c++, modify them there and then pass back. The data structures map without problems so I propose we have all defined on the fortran side. See attached
> [example.zip](https://github.com/madgraph5/madgraph4gpu/files/6399880/example.zip)

Hi @roiser sorry I had missed your message... that's great! Indeed, much much simpler than I thought!
And your fortran skills are now better than mine :-)

Impressive that even complex passing to std::complex works so easily out of the box. I am not sure we will need that, but that's great to know.

Very good then, I agree with you, we can define the input and output arrays on the fortran side. We just need to be sure that we get the total size right, but also the AOSOA interpretation. One suggestion, can you try to play with a multidimensional double array and see how easy it is to get it right across Fortran and C++?

Essentially what we will need to pass across are the three arrays that currently in check.cc are here, momenta, MEs and weights (and I am not even 100% convinced we need to pass weights... this is only if we want to vectorize in c++ a few multiplications, but it may be usefl to pass them anyway):
```
  const int nMomenta = np4*npar*nevt; // (NB: nevt=npagM*neppM for AOSOA layouts)
  const int nWeights = nevt;
  const int nMEs     = nevt; // FIXME: assume process.nprocesses == 1 (eventually: nMEs = nevt * nprocesses?)
  auto hstMomenta   = hstMakeUnique<fptype_sv>( nMomenta ); // AOSOA[npagM][npar][np4][neppM] (NB: nevt=npagM*neppM)
  auto hstWeights   = hstMakeUnique<fptype   >( nWeights );
  auto hstMEs       = hstMakeUnique<fptype_sv>( nMEs ); // AOSOA[npagM][neppM] (NB: nevt=npagM*neppM)
```
Currently only momenta are an AOSOA, while weights and MEs are 1d arrays. But this is for eemumu, where we have a single process, and also where we are not really using the weights to do anything. @oliviermattelaer knows better what we need to do with nprocesses>1 (is it only for pp and/or NLO?).

Anyway: I would suggest to play with something that looks like the momenta AOSOA[npagM][npar][np4][neppM]. Does this sound reasonable?

**Comment by valassi** [comment]
PS then we need to make sure that we encapsulate (probably in CPPProcess) a single C++ (C?) function that takes care of dispatching either to vectorized C++ or to the cuda kernel.

Hm on second thought there is however one parameter that we must pass from c++ to Fortran in advance, namely the nevt number of events (ie the dimension of the grid in CUDA)...? Or maybe it's vicevresa, since the main is Fortran, it is going to be Fortran that decides the grid size, either hardcoded or taking a command line argument (later on).

**Comment by oliviermattelaer** [comment]
Another issue that I realize now is the fact that we have multiple matrix-element computation possible in a single executable but rather than calling all of them, we typically (actually always) Monte-Carlo overthem.
This is problematic since in general 2 consecutive event will not evaluate the same matrix-element function.
(for q q~ > w+ w- q q~ we have 32 different matrix-element function).

If number of event per page is small compare to the number of event in the iteration, then one solution is to force all event of a given page to use the same matrix-element (or to have a page per matrix-element and wait for a page to be full to launch the computation). But this will not scale for GPU...

Note that such issue only exists at LO. So it might not be too crazy to forbid those monte-carlo approach and then integrate a single one at the time.

**Comment by valassi** [comment]
Hi @oliviermattelaer thanks. 

I am not completely sure I visualize the problem. I think that having some code would be easier to understand it, maybe I will try to generate a gridpack, you say q q~ > w+ w- q q~ for instance?

There is one issue which I know we (at least, I) have completely neglected so far, the "nprocesses" which is assumed equal to 1 in many parts of the code (with big "FIXME" written all around them). Is this one example of this, or is this still a different problem?

Similarly, I know the Fortran code I have seen has "matrix1_", but I guess you can have some with "matrix2_" as well, etc. Again, is this pointing in the same direction, or is it another problem?

I do not understand why this is only at LO actually? Do you not have it if you do q q~ > w+ w- q q~ g at NLO for instance?

Anyway: I may be completely wrong, but my take is that this is completely orthogonal. In any case for any process we will always have several (much greater than 1!) different event kinematics to be computed (ie events with the same exact particle types, but with different momenta). What I had in mind is that we compute some of those events in parallel. As long as we do the bookkeeping right of inputs and outputs, I do not see a problem. At least, vectorization, with 4 or maybe 8 or maybe 16 events in parallel, should not be a problem. With GPU the issue is that typically we need thousands (hundreds of thousands?) of events in parallel to be efficient, see the throughput plots by Stefan and Andy... but this is for every single process, and independent of Fortran linking. This is not necessarily a problem, as long as we have enough memory on the CPU (and GPU?) to keep the data, and as a long as the preprocessing and postprocessing (eg saving to file) per event is not too time consuming.

**Comment by oliviermattelaer** [comment]
The issue that I point here is the one of multiple matrix_X.f one 
For a given phase-space point only one of them is evaluated. (in a random way relative to the PDF contribution)

> I do not understand why this is only at LO actually? Do you not have it if you do q q~ > w+ w- q q~ g at NLO for instance?

Just that NLO does not optimise on that and therefore you never have any matrix2.f (but has similar complication that consecutive phase-space point might not need the same matrix-element evaluation since you have various counter-term).

> Anyway: I may be completely wrong, but my take is that this is completely orthogonal. 

In madevent, the first iteration is typically done with quite few evaluation of the phase-space. (the default is 1000 for the survey and up to 5000 in the refine stage). So at some point we might need to change those number for GPU
